crates/engine/tree/src/tree/mod.rs

```diff
@@ -2650,6 +2663,245 @@ where
+
+        // after executing the block we can stop executing transactions
+        handle.stop_prewarming_execution();
+
+        if let Err(err) = self.consensus.validate_block_post_execution(&block, &output) {
+            // call post-block hook
+            self.invalid_block_hook.on_invalid_block(&parent_block, &block, &output, None);
+            return Err(err.into())
+        }
+
+        let hashed_state = self.provider.hashed_post_state(&output.state);
+
+        if let Err(err) = self
+            .payload_validator
+            .validate_block_post_execution_with_hashed_state(&hashed_state, &block)
+        {
+            // call post-block hook
+            self.invalid_block_hook.on_invalid_block(&parent_block, &block, &output, None);
+            return Err(err.into())
+        }
+
+        trace!(target: "engine::tree", block=?block_num_hash, "Calculating block state root");
+
+        let root_time = Instant::now();
+
+        //  (state_root, trie_output, root_elapsed)
+        let mut maybe_state_root = None;
+
+        if is_descendant_of_persisting_blocks {
+            // if we new payload extends the current canonical change we attempt to use the
+            // background task or try to compute it in parallel
+            if self.config.use_state_root_task() {
+                match handle.state_root() {
+                    Ok(res) => {
+                        // we double check the state root here
+                        // TODO: clean this ups
+                        if res.state_root.0 == block.header().state_root() {
+                            maybe_state_root =
+                                Some((res.state_root.0, res.state_root.1, res.total_time))
+                        }
+                    }
+                    Err(err) => {
+                        // TODO error handling
+                    }
+                }
+            } else {
+                match self.compute_state_root_parallel(block.header().parent_hash(), &hashed_state)
+                {
+                    Ok(result) => {
+                        info!(
+                            target: "engine::tree",
+                            block = ?block_num_hash,
+                            regular_state_root = ?result.0,
+                            "Regular root task finished"
+                        );
+                        maybe_state_root = Some((result.0, result.1, root_time.elapsed()));
+                    }
+                    Err(ParallelStateRootError::Provider(ProviderError::ConsistentView(error))) => {
+                        debug!(target: "engine", %error, "Parallel state root computation failed consistency check, falling back");
+                        let (root, updates) =
+                            state_provider.state_root_with_updates(hashed_state.clone())?;
+                        maybe_state_root = Some((root, updates, root_time.elapsed()));
```

Comment:
this can be removed, because we will fall back to calling `state_provider.state_root_with_updates` below anyway